File: frontend/src/utils/performanceMonitor.ts
/**
 * Frontend performance monitoring system for authentication flows.
 * 
 * This module provides comprehensive performance monitoring, optimization utilities,
 * and caching strategies for the frontend authentication system.
 */

import { useState, useEffect, useCallback, useRef } from 'react';

export interface PerformanceMetric {
  operation: string;
  duration: number;
  timestamp: number;
  success: boolean;
  metadata?: Record<string, any>;
}

export interface CacheEntry<T = any> {
  value: T;
  timestamp: number;
  ttl: number;
  hits: number;
}

export interface PerformanceReport {
  metrics: Record<string, PerformanceMetric[]>;
  cacheStats: {
    totalEntries: number;
    hitRate: number;
    totalHits: number;
    totalMisses: number;
  };
  slowOperations: PerformanceMetric[];
  recommendations: string[];
}

class FrontendPerformanceMonitor {
  private metrics: Record<string, PerformanceMetric[]> = {};
  private cache: Map<string, CacheEntry> = new Map();
  private cacheHits = 0;
  private cacheMisses = 0;
  private observers: Set<(report: PerformanceReport) => void> = new Set();

  // Record a performance metric
  recordMetric(
    operation: string,
    duration: number,
    success: boolean = true,
    metadata?: Record<string, any>
  ): void {
    const metric: PerformanceMetric = {
      operation,
      duration,
      timestamp: Date.now(),
      success,
      metadata
    };

    if (!this.metrics[operation]) {
      this.metrics[operation] = [];
    }

    this.metrics[operation].push(metric);

    // Keep only last 100 metrics per operation
    if (this.metrics[operation].length > 100) {
      this.metrics[operation] = this.metrics[operation].slice(-100);
    }

    // Log slow operations (> 1 second)
    if (duration > 1000) {
      console.warn(`Slow operation detected: ${operation} took ${duration}ms`);
    }

    this.notifyObservers();
  }

  // Measure and record an async operation
  async measureAsync<T>(
    operation: string,
    fn: () => Promise<T>,
    metadata?: Record<string, any>
  ): Promise<T> {
    const startTime = performance.now();
    let success = true;
    let error: Error | null = null;

    try {
      const result = await fn();
      return result;
    } catch (e) {
      success = false;
      error = e as Error;
      throw e;
    } finally {
      const duration = performance.now() - startTime;
      this.recordMetric(operation, duration, success, {
        ...metadata,
        error: error?.message
      });
    }
  }

  // Measure and record a sync operation
  measureSync<T>(
    operation: string,
    fn: () => T,
    metadata?: Record<string, any>
  ): T {
    const startTime = performance.now();
    let success = true;
    let error: Error | null = null;

    try {
      const result = fn();
      return result;
    } catch (e) {
      success = false;
      error = e as Error;
      throw e;
    } finally {
      const duration = performance.now() - startTime;
      this.recordMetric(operation, duration, success, {
        ...metadata,
        error: error?.message
      });
    }
  }

  // Cache management
  setCache<T>(key: string, value: T, ttl: number = 300000): void { // 5 minutes default
    this.cache.set(key, {
      value,
      timestamp: Date.now(),
      ttl,
      hits: 0
    });
  }

  getCache<T>(key: string): T | null {
    const entry = this.cache.get(key);
    
    if (!entry) {
      this.cacheMisses++;
      return null;
    }

    // Check if expired
    if (Date.now() - entry.timestamp > entry.ttl) {
      this.cache.delete(key);
      this.cacheMisses++;
      return null;
    }

    entry.hits++;
    this.cacheHits++;
    return entry.value as T;
  }

  clearCache(): void {
    this.cache.clear();
  }

  // Get performance statistics
  getOperationStats(operation: string): {
    totalCalls: number;
    successRate: number;
    avgDuration: number;
    minDuration: number;
    maxDuration: number;
    p95Duration: number;
  } | null {
    const metrics = this.metrics[operation];
    if (!metrics || metrics.length === 0) {
      return null;
    }

    const durations = metrics.map(m => m.duration);
    const successes = metrics.filter(m => m.success).length;

    return {
      totalCalls: metrics.length,
      successRate: (successes / metrics.length) * 100,
      avgDuration: durations.reduce((a, b) => a + b, 0) / durations.length,
      minDuration: Math.min(...durations),
      maxDuration: Math.max(...durations),
      p95Duration: this.percentile(durations, 95)
    };
  }

  // Get slow operations (> 1 second)
  getSlowOperations(limit: number = 20): PerformanceMetric[] {
    const allMetrics = Object.values(this.metrics).flat();
    return allMetrics
      .filter(m => m.duration > 1000)
      .sort((a, b) => b.duration - a.duration)
      .slice(0, limit);
  }

  // Generate performance report
  getPerformanceReport(): PerformanceReport {
    const totalCacheRequests = this.cacheHits + this.cacheMisses;
    
    return {
      metrics: { ...this.metrics },
      cacheStats: {
        totalEntries: this.cache.size,
        hitRate: totalCacheRequests > 0 ? (this.cacheHits / totalCacheRequests) * 100 : 0,
        totalHits: this.cacheHits,
        totalMisses: this.cacheMisses
      },
      slowOperations: this.getSlowOperations(),
      recommendations: this.generateRecommendations()
    };
  }

  // Generate performance recommendations
  private generateRecommendations(): string[] {
    const recommendations: string[] = [];
    const slowOps = this.getSlowOperations(5);
    
    if (slowOps.length > 0) {
      recommendations.push(
        `Found ${slowOps.length} slow operations. Consider optimizing: ${slowOps.map(op => op.operation).join(', ')}`
      );
    }

    const totalCacheRequests = this.cacheHits + this.cacheMisses;
    if (totalCacheRequests > 0) {
      const hitRate = (this.cacheHits / totalCacheRequests) * 100;
      if (hitRate < 70) {
        recommendations.push(
          `Cache hit rate is ${hitRate.toFixed(1)}%. Consider increasing cache TTL or improving cache strategy.`
        );
      }
    }

    // Check for operations with high error rates
    Object.entries(this.metrics).forEach(([operation, metrics]) => {
      const successRate = (metrics.filter(m => m.success).length / metrics.length) * 100;
      if (successRate < 95 && metrics.length > 5) {
        recommendations.push(
          `Operation '${operation}' has ${successRate.toFixed(1)}% success rate. Investigate errors.`
        );
      }
    });

    return recommendations;
  }

  // Subscribe to performance updates
  subscribe(observer: (report: PerformanceReport) => void): () => void {
    this.observers.add(observer);
    return () => this.observers.delete(observer);
  }

  private notifyObservers(): void {
    const report = this.getPerformanceReport();
    this.observers.forEach(observer => observer(report));
  }

  private percentile(arr: number[], p: number): number {
    const sorted = [...arr].sort((a, b) => a - b);
    const index = Math.ceil((p / 100) * sorted.length) - 1;
    return sorted[Math.max(0, Math.min(index, sorted.length - 1))];
  }

  // Cleanup expired cache entries
  cleanupCache(): number {
    const now = Date.now();
    let removed = 0;

    for (const [key, entry] of this.cache.entries()) {
      if (now - entry.timestamp > entry.ttl) {
        this.cache.delete(key);
        removed++;
      }
    }

    return removed;
  }
}

// Global performance monitor instance
export const performanceMonitor = new FrontendPerformanceMonitor();

// React hooks for performance monitoring
export function usePerformanceMonitor(operation?: string) {
  const [report, setReport] = useState<PerformanceReport | null>(null);

  useEffect(() => {
    const unsubscribe = performanceMonitor.subscribe(setReport);
    return unsubscribe;
  }, []);

  const measureAsync = useCallback(async <T>(
    fn: () => Promise<T>,
    operationName?: string,
    metadata?: Record<string, any>
  ): Promise<T> => {
    const opName = operationName || operation || 'unknown';
    return performanceMonitor.measureAsync(opName, fn, metadata);
  }, [operation]);

  const measureSync = useCallback(<T>(
    fn: () => T,
    operationName?: string,
    metadata?: Record<string, any>
  ): T => {
    const opName = operationName || operation || 'unknown';
    return performanceMonitor.measureSync(opName, fn, metadata);
  }, [operation]);

  const recordMetric = useCallback((
    duration: number,
    success: boolean = true,
    operationName?: string,
    metadata?: Record<string, any>
  ) => {
    const opName = operationName || operation || 'unknown';
    performanceMonitor.recordMetric(opName, duration, success, metadata);
  }, [operation]);

  return {
    report,
    measureAsync,
    measureSync,
    recordMetric
  };
}

// Higher-order component for automatic performance monitoring
export function withPerformanceMonitoring<T extends object>(
  Component: React.ComponentType<T>,
  operationName: string
): React.ComponentType<T> {
  return function PerformanceMonitoredComponent(props: T) {
    const renderStartTime = useRef(performance.now());
    
    useEffect(() => {
      const renderDuration = performance.now() - renderStartTime.current;
      performanceMonitor.recordMetric(`${operationName}_render`, renderDuration);
    });

    return <Component {...props} />;
  };
}

// Optimized fetch wrapper with caching and performance monitoring
export async function optimizedFetch(
  url: string,
  options: RequestInit = {},
  cacheKey?: string,
  cacheTtl: number = 300000
): Promise<Response> {
  // Try cache first if cache key provided
  if (cacheKey) {
    const cached = performanceMonitor.getCache<Response>(cacheKey);
    if (cached) {
      return cached.clone();
    }
  }

  // Perform fetch with monitoring
  const response = await performanceMonitor.measureAsync(
    'fetch',
    async () => {
      const resp = await fetch(url, options);
      
      // Cache successful responses
      if (cacheKey && resp.ok) {
        performanceMonitor.setCache(cacheKey, resp.clone(), cacheTtl);
      }
      
      return resp;
    },
    { url, method: options.method || 'GET' }
  );

  return response;
}

// Authentication-specific performance utilities
export class AuthPerformanceUtils {
  // Optimized token validation with caching
  static async validateTokenCached(token: string): Promise<boolean> {
    const cacheKey = `token_valid_${token.slice(-10)}`;
    
    const cached = performanceMonitor.getCache<boolean>(cacheKey);
    if (cached !== null) {
      return cached;
    }

    const isValid = await performanceMonitor.measureAsync(
      'token_validation',
      async () => {
        const response = await fetch('/api/v1/auth/verify', {
          headers: { Authorization: `Bearer ${token}` }
        });
        return response.ok;
      }
    );

    // Cache for 5 minutes
    performanceMonitor.setCache(cacheKey, isValid, 300000);
    return isValid;
  }

  // Optimized user profile loading with caching
  static async loadUserProfileCached(userId: string): Promise<any> {
    const cacheKey = `user_profile_${userId}`;
    
    const cached = performanceMonitor.getCache(cacheKey);
    if (cached) {
      return cached;
    }

    const profile = await performanceMonitor.measureAsync(
      'load_user_profile',
      async () => {
        const response = await fetch('/api/v1/users/me', {
          headers: { Authorization: `Bearer ${localStorage.getItem('auth_token')}` }
        });
        return response.json();
      }
    );

    // Cache for 10 minutes
    performanceMonitor.setCache(cacheKey, profile, 600000);
    return profile;
  }

  // Debounced profile update to prevent excessive API calls
  static debounceProfileUpdate = (() => {
    let timeoutId: NodeJS.Timeout;
    
    return (updateData: any, delay: number = 1000): Promise<void> => {
      return new Promise((resolve, reject) => {
        clearTimeout(timeoutId);
        
        timeoutId = setTimeout(async () => {
          try {
            await performanceMonitor.measureAsync(
              'update_user_profile',
              async () => {
                const response = await fetch('/api/v1/users/profile', {
                  method: 'PUT',
                  headers: {
                    'Content-Type': 'application/json',
                    Authorization: `Bearer ${localStorage.getItem('auth_token')}`
                  },
                  body: JSON.stringify(updateData)
                });
                
                if (!response.ok) {
                  throw new Error('Profile update failed');
                }
                
                return response.json();
              }
            );
            
            // Invalidate cache
            performanceMonitor.clearCache();
            resolve();
          } catch (error) {
            reject(error);
          }
        }, delay);
      });
    };
  })();
}

// Performance monitoring initialization
export function initializePerformanceMonitoring(): void {
  // Set up periodic cache cleanup
  setInterval(() => {
    const removed = performanceMonitor.cleanupCache();
    if (removed > 0) {
      console.log(`Cleaned up ${removed} expired cache entries`);
    }
  }, 60000); // Every minute

  // Monitor page load performance
  if (typeof window !== 'undefined') {
    window.addEventListener('load', () => {
      const navigation = performance.getEntriesByType('navigation')[0] as PerformanceNavigationTiming;
      if (navigation) {
        performanceMonitor.recordMetric(
          'page_load',
          navigation.loadEventEnd - navigation.fetchStart,
          true,
          {
            domContentLoaded: navigation.domContentLoadedEventEnd - navigation.fetchStart,
            firstPaint: navigation.responseEnd - navigation.fetchStart
          }
        );
      }
    });

    // Monitor route changes (for SPAs)
    let lastUrl = location.href;
    new MutationObserver(() => {
      const url = location.href;
      if (url !== lastUrl) {
        performanceMonitor.recordMetric('route_change', performance.now(), true, { url });
        lastUrl = url;
      }
    }).observe(document, { subtree: true, childList: true });
  }

  console.log('Frontend performance monitoring initialized');
}

// Export performance report for debugging
export function logPerformanceReport(): void {
  const report = performanceMonitor.getPerformanceReport();
  console.group('🚀 Performance Report');
  console.log('Metrics:', report.metrics);
  console.log('Cache Stats:', report.cacheStats);
  console.log('Slow Operations:', report.slowOperations);
  console.log('Recommendations:', report.recommendations);
  console.groupEnd();
}

File: backend/src/utils/performance_monitor.py
"""
Performance monitoring system for authentication flows.

This module provides comprehensive performance monitoring, caching strategies,
and optimization utilities for the authentication system.
"""

import time
import logging
import asyncio
from typing import Dict, Any, Optional, List, Callable
from datetime import datetime, timedelta
from functools import wraps
from contextlib import asynccontextmanager
import json
import hashlib

logger = logging.getLogger(__name__)


class PerformanceMetrics:
    """Collects and manages performance metrics."""
    
    def __init__(self):
        self.metrics: Dict[str, List[Dict[str, Any]]] = {}
        self.cache_stats: Dict[str, Dict[str, int]] = {}
        self.slow_operations: List[Dict[str, Any]] = []
        
    def record_operation(
        self, 
        operation: str, 
        duration: float, 
        success: bool = True,
        metadata: Optional[Dict[str, Any]] = None
    ):
        """Record a performance metric for an operation."""
        if operation not in self.metrics:
            self.metrics[operation] = []
            
        metric = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "duration": duration,
            "success": success,
            "metadata": metadata or {}
        }
        
        self.metrics[operation].append(metric)
        
        # Track slow operations (> 2 seconds)
        if duration > 2.0:
            self.slow_operations.append({
                "operation": operation,
                "duration": duration,
                "timestamp": metric["timestamp"],
                "metadata": metadata
            })
            logger.warning(f"Slow operation detected: {operation} took {duration:.2f}s")
        
        # Keep only last 1000 metrics per operation
        if len(self.metrics[operation]) > 1000:
            self.metrics[operation] = self.metrics[operation][-1000:]
    
    def record_cache_hit(self, cache_key: str):
        """Record a cache hit."""
        if cache_key not in self.cache_stats:
            self.cache_stats[cache_key] = {"hits": 0, "misses": 0}
        self.cache_stats[cache_key]["hits"] += 1
    
    def record_cache_miss(self, cache_key: str):
        """Record a cache miss."""
        if cache_key not in self.cache_stats:
            self.cache_stats[cache_key] = {"hits": 0, "misses": 0}
        self.cache_stats[cache_key]["misses"] += 1
    
    def get_operation_stats(self, operation: str) -> Dict[str, Any]:
        """Get statistics for a specific operation."""
        if operation not in self.metrics:
            return {"error": "Operation not found"}
        
        durations = [m["duration"] for m in self.metrics[operation]]
        successes = [m["success"] for m in self.metrics[operation]]
        
        if not durations:
            return {"error": "No data available"}
        
        return {
            "operation": operation,
            "total_calls": len(durations),
            "success_rate": sum(successes) / len(successes) * 100,
            "avg_duration": sum(durations) / len(durations),
            "min_duration": min(durations),
            "max_duration": max(durations),
            "p95_duration": self._percentile(durations, 95),
            "p99_duration": self._percentile(durations, 99),
            "recent_calls": self.metrics[operation][-10:]  # Last 10 calls
        }
    
    def get_cache_stats(self) -> Dict[str, Any]:
        """Get cache statistics."""
        total_hits = sum(stats["hits"] for stats in self.cache_stats.values())
        total_misses = sum(stats["misses"] for stats in self.cache_stats.values())
        total_requests = total_hits + total_misses
        
        return {
            "total_requests": total_requests,
            "total_hits": total_hits,
            "total_misses": total_misses,
            "hit_rate": (total_hits / total_requests * 100) if total_requests > 0 else 0,
            "cache_details": self.cache_stats
        }
    
    def get_slow_operations(self, limit: int = 50) -> List[Dict[str, Any]]:
        """Get recent slow operations."""
        return sorted(
            self.slow_operations[-limit:], 
            key=lambda x: x["duration"], 
            reverse=True
        )
    
    def get_summary(self) -> Dict[str, Any]:
        """Get a summary of all performance metrics."""
        summary = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "operations": {},
            "cache_stats": self.get_cache_stats(),
            "slow_operations_count": len(self.slow_operations),
            "total_operations": sum(len(metrics) for metrics in self.metrics.values())
        }
        
        for operation in self.metrics.keys():
            summary["operations"][operation] = self.get_operation_stats(operation)
        
        return summary
    
    @staticmethod
    def _percentile(data: List[float], percentile: int) -> float:
        """Calculate percentile of a list of numbers."""
        if not data:
            return 0.0
        sorted_data = sorted(data)
        index = int(len(sorted_data) * percentile / 100)
        return sorted_data[min(index, len(sorted_data) - 1)]


class CacheManager:
    """In-memory cache manager with TTL support."""
    
    def __init__(self, default_ttl: int = 300):  # 5 minutes default
        self.cache: Dict[str, Dict[str, Any]] = {}
        self.default_ttl = default_ttl
        
    def get(self, key: str) -> Optional[Any]:
        """Get value from cache."""
        if key in self.cache:
            entry = self.cache[key]
            if datetime.utcnow() < entry["expires"]:
                performance_metrics.record_cache_hit(key)
                return entry["value"]
            else:
                # Expired, remove from cache
                del self.cache[key]
        
        performance_metrics.record_cache_miss(key)
        return None
    
    def set(self, key: str, value: Any, ttl: Optional[int] = None) -> None:
        """Set value in cache with TTL."""
        ttl = ttl or self.default_ttl
        expires = datetime.utcnow() + timedelta(seconds=ttl)
        
        self.cache[key] = {
            "value": value,
            "expires": expires,
            "created": datetime.utcnow()
        }
    
    def delete(self, key: str) -> bool:
        """Delete key from cache."""
        if key in self.cache:
            del self.cache[key]
            return True
        return False
    
    def clear(self) -> None:
        """Clear all cache entries."""
        self.cache.clear()
    
    def cleanup_expired(self) -> int:
        """Remove expired entries and return count of removed items."""
        now = datetime.utcnow()
        expired_keys = [
            key for key, entry in self.cache.items() 
            if now >= entry["expires"]
        ]
        
        for key in expired_keys:
            del self.cache[key]
        
        return len(expired_keys)
    
    def get_stats(self) -> Dict[str, Any]:
        """Get cache statistics."""
        now = datetime.utcnow()
        active_entries = sum(
            1 for entry in self.cache.values() 
            if now < entry["expires"]
        )
        
        return {
            "total_entries": len(self.cache),
            "active_entries": active_entries,
            "expired_entries": len(self.cache) - active_entries,
            "memory_usage_estimate": len(str(self.cache))  # Rough estimate
        }


# Global instances
performance_metrics = PerformanceMetrics()
cache_manager = CacheManager()


def monitor_performance(operation_name: str):
    """Decorator to monitor function performance."""
    def decorator(func):
        if asyncio.iscoroutinefunction(func):
            @wraps(func)
            async def async_wrapper(*args, **kwargs):
                start_time = time.time()
                success = True
                error = None
                
                try:
                    result = await func(*args, **kwargs)
                    return result
                except Exception as e:
                    success = False
                    error = str(e)
                    raise
                finally:
                    duration = time.time() - start_time
                    performance_metrics.record_operation(
                        operation_name, 
                        duration, 
                        success,
                        {"error": error} if error else None
                    )
            
            return async_wrapper
        else:
            @wraps(func)
            def sync_wrapper(*args, **kwargs):
                start_time = time.time()
                success = True
                error = None
                
                try:
                    result = func(*args, **kwargs)
                    return result
                except Exception as e:
                    success = False
                    error = str(e)
                    raise
                finally:
                    duration = time.time() - start_time
                    performance_metrics.record_operation(
                        operation_name, 
                        duration, 
                        success,
                        {"error": error} if error else None
                    )
            
            return sync_wrapper
    
    return decorator


@asynccontextmanager
async def performance_context(operation_name: str, metadata: Optional[Dict[str, Any]] = None):
    """Context manager for monitoring performance."""
    start_time = time.time()
    success = True
    error = None
    
    try:
        yield
    except Exception as e:
        success = False
        error = str(e)
        raise
    finally:
        duration = time.time() - start_time
        final_metadata = metadata or {}
        if error:
            final_metadata["error"] = error
        
        performance_metrics.record_operation(
            operation_name, 
            duration, 
            success,
            final_metadata
        )


def cached(key_func: Callable = None, ttl: int = 300):
    """Decorator to cache function results."""
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            # Generate cache key
            if key_func:
                cache_key = key_func(*args, **kwargs)
            else:
                # Default key generation
                key_parts = [func.__name__]
                key_parts.extend(str(arg) for arg in args)
                key_parts.extend(f"{k}={v}" for k, v in sorted(kwargs.items()))
                cache_key = hashlib.md5("|".join(key_parts).encode()).hexdigest()
            
            # Try to get from cache
            cached_result = cache_manager.get(cache_key)
            if cached_result is not None:
                return cached_result
            
            # Execute function and cache result
            result = await func(*args, **kwargs)
            cache_manager.set(cache_key, result, ttl)
            
            return result
        
        return wrapper
    return decorator


class AuthPerformanceOptimizer:
    """Authentication-specific performance optimizations."""
    
    @staticmethod
    @monitor_performance("password_hash")
    def optimize_password_hashing(password: str) -> str:
        """Optimized password hashing with performance monitoring."""
        from ..services.password_service import password_service
        return password_service.hash_password(password)
    
    @staticmethod
    @monitor_performance("jwt_create")
    def optimize_jwt_creation(payload: Dict[str, Any]) -> str:
        """Optimized JWT creation with performance monitoring."""
        from ..services.jwt_service import jwt_service
        return jwt_service.create_token(payload)
    
    @staticmethod
    @monitor_performance("jwt_verify")
    def optimize_jwt_verification(token: str) -> Dict[str, Any]:
        """Optimized JWT verification with performance monitoring."""
        from ..services.jwt_service import jwt_service
        return jwt_service.verify_token(token)
    
    @staticmethod
    @cached(ttl=300)  # Cache user data for 5 minutes
    @monitor_performance("user_lookup")
    async def optimize_user_lookup(user_id: str) -> Optional[Dict[str, Any]]:
        """Optimized user lookup with caching."""
        from ..services.dynamodb import db_service
        return await db_service.get_user_by_id(user_id)
    
    @staticmethod
    @cached(ttl=600)  # Cache user by email for 10 minutes
    @monitor_performance("user_email_lookup")
    async def optimize_user_email_lookup(email: str) -> Optional[Dict[str, Any]]:
        """Optimized user email lookup with caching."""
        from ..services.dynamodb import db_service
        return await db_service.get_user_by_email(email)


# Background task to cleanup expired cache entries
async def cache_cleanup_task():
    """Background task to cleanup expired cache entries."""
    while True:
        try:
            removed_count = cache_manager.cleanup_expired()
            if removed_count > 0:
                logger.info(f"Cleaned up {removed_count} expired cache entries")
        except Exception as e:
            logger.error(f"Cache cleanup failed: {e}")
        
        # Run cleanup every 5 minutes
        await asyncio.sleep(300)


# Performance monitoring endpoints
def get_performance_report() -> Dict[str, Any]:
    """Get comprehensive performance report."""
    return {
        "performance_metrics": performance_metrics.get_summary(),
        "cache_stats": cache_manager.get_stats(),
        "slow_operations": performance_metrics.get_slow_operations(20),
        "recommendations": generate_performance_recommendations()
    }


def generate_performance_recommendations() -> List[Dict[str, str]]:
    """Generate performance optimization recommendations."""
    recommendations = []
    
    # Check for slow operations
    slow_ops = performance_metrics.get_slow_operations(10)
    if slow_ops:
        recommendations.append({
            "type": "slow_operations",
            "message": f"Found {len(slow_ops)} slow operations. Consider optimizing these endpoints.",
            "action": "Review slow operations and implement caching or optimization"
        })
    
    # Check cache hit rate
    cache_stats = cache_manager.get_stats()
    if cache_stats["total_entries"] > 0:
        cache_perf_stats = performance_metrics.get_cache_stats()
        if cache_perf_stats["hit_rate"] < 70:
            recommendations.append({
                "type": "low_cache_hit_rate",
                "message": f"Cache hit rate is {cache_perf_stats['hit_rate']:.1f}%. Consider increasing TTL or improving cache keys.",
                "action": "Analyze cache usage patterns and optimize caching strategy"
            })
    
    # Check for high error rates
    for operation, stats in performance_metrics.get_summary()["operations"].items():
        if stats.get("success_rate", 100) < 95:
            recommendations.append({
                "type": "high_error_rate",
                "message": f"Operation '{operation}' has {stats['success_rate']:.1f}% success rate.",
                "action": "Investigate and fix errors in this operation"
            })
    
    return recommendations


# Startup function to initialize performance monitoring
def initialize_performance_monitoring():
    """Initialize performance monitoring system."""
    logger.info("Initializing performance monitoring system...")
    
    # Start cache cleanup task
    asyncio.create_task(cache_cleanup_task())
    
    logger.info("Performance monitoring system initialized")


# Export optimized functions for use in authentication flows
auth_optimizer = AuthPerformanceOptimizer()